### DISCUSSION TITLE
pb_hooks why have multiple *.pb.js files?

### DISCUSSION THREAD

**Discussion by VictorioBerra** [pocketbase/pocketbase#4561] [Q&A]
The docs say:

> You can start by creating *.pb.js file(s) inside a pb_hooks directory next to your executable.

Why would I want multiple files? Why not just put everything in one `main.pb.js`? If i want to break up my code, the docs say I can `require(`${__hooks}/utils.js`)`. 

The docs also say:

> Each handler function (hook, route, middleware, etc.) is serialized and executed in its own isolated context as a separate "program"

So if I have a single `main.pb.js` all my handlers inside are already treated as a separate "program". Why break them out into their own file?

**Comment by ganigeorgiev** [comment]
> Why would I want multiple files? 

I don't know how to answer this. You can ask yourself the same question about your other regular application files. It is up to the users to decide whether they'll put all of their hooks in 1 big file or split them into several. The same apply if you use Go.

> If i want to break up my code, the docs say I can require(${__hooks}/utils.js).

Not exactly. You can use `require` if your hook handlers share a piece of code, but the hooks need to be invoked at the top level since hooks are expected to be registered only once. For more information how the JSVM works you can refer to https://github.com/pocketbase/pocketbase/discussions/2825#discussioncomment-6376210.

**Reply by ganigeorgiev** [reply]
_Oh, I actually forgot that `require` will cache the file in the registry after the first load, so it should work even if you put your hooks in regular `.js` files and then require them inside a `.pb.js` (_it should be safe even if required multiple times as duplicated calls will do nothing; but this in general is more of a side-effect and works only because `require` is enabled for the loader runtime_)._

**Reply by VictorioBerra** [reply]
@ganigeorgiev Thanks for that insight. I actually did not mean to suggest id require hooks from regular JS files but that is good to know to avoid.
